Migrate some kill integration cli test to api tests

vdemeester opened the pull request:
Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented on client/client.go:
```diff
@@ -107,7 +107,7 @@ func CheckRedirect(req *http.Request, via []*http.Request) error {
 // Use DOCKER_CERT_PATH to load the TLS certificates from.
 // Use DOCKER_TLS_VERIFY to enable or disable TLS verification, off by default.
-func NewEnvClient() (*Client, error) {
+func NewEnvClient(ops ...func(*Client)) (*Client, error) {
```
These changes on `client` are not strictly required (and might be better off discussed in a separate PR)

dnephin replied:
I think we've wanted to make this change for a while. What do you think about something like this:

* create a `func FromEnv(*Client)` and move all the code from `NewEnvClient` (except the `ops` addition) to `FromEnv()`
* create `NewClientWithOpts()` (or something like that) which creates a client from ops, but has no other inline setup.
* remove `ops`  from `NewEnvClient` and deprecate it. Replace the implementation with `NewClientWithOpts(FromEnv)`.

That way the new interface would create a bare `client{}` and we can apply any ops onto it. If there are required fields (host and client?) maybe it would also validate that.

vdemeester replied:
sounds very good to me 👼

vdemeester commented on integration-cli/docker_cli_kill_test.go:
```diff
@@ -6,37 +6,9 @@ import (
-	})
-}
-
 func (s *DockerSuite) TestKillDifferentUserContainer(c *check.C) {
```
This one shouldn't be hard to port either

dnephin commented on integration/container/kill_test.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+func TestKillWithStopSignalAndRestartPolicies(t *testing.T) {
+	skip.If(t, testEnv.OSType != "linux") // Windows only supports 1.25 or later
```
Instead of a comment this can be the last arg to `slip.If()` so that we see it in the test output. (same below).

vdemeester replied:
good point 🙃

vdemeester committed:
Introduce NewClientWithOpts func to build custom client easily

This allows to create a client with default values and override those
using functors. As an example, `NewEnvClient()` becomes
`NewClientWithOpts(FromEnv)` ; and if you want a different api version
for this client : `NewClientWithOpts(FromEnv, WithVersion("1.35"))`

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Move some kill integration cli to api tests

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester committed:
Migrate some calls to new client function

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

dnephin commented on integration/container/kill_test.go:
```diff
@@ -0,0 +1,197 @@
+	require.NoError(t, err)
+}
+
+func containerStatusIs(ctx context.Context, client client.APIClient, containerID, status string) func(log poll.LogT) poll.Result {
```
I think this already exists as `containerIsInState` in this package.

dnephin reviewed (commented):
looks like lint failures, and one comment, otherwise LGTM

dnephin reviewed (approved):
LGTM

dnephin commented on integration/network/inspect_test.go:
```diff
@@ -24,7 +23,7 @@ func TestInspectNetwork(t *testing.T) {
 	d := newSwarm(t)
 	defer d.Stop(t)
-	client, err := request.NewClientForHost(d.Sock())
+	client, err := client.NewClientWithOpts(client.WithHost((d.Sock())))
```
looks like this test failed

> Cannot connect to the Docker daemon at unix:///tmp/docker-integration/df6a1d02502f7.sock

cpuguy83 commented on integration/container/kill_test.go:
```diff
@@ -0,0 +1,182 @@
+)
+
+func TestKillContainerInvalidSignal(t *testing.T) {
+	defer setupTest(t)()
```
I don't think this works with `t.Parallel()`
For what it's worth, the auto-cleanup is probably not needed as it's simple to cleanup manually.

dnephin replied:
Good catch. Altough we don't actually use `-parallel` right now so these `t.Parallel()` don't do anything I believe.

cpuguy83 replied:
I think we should make the extra effort to have these run parallel correctly, as we move stuff over it will be a huge help to test run times.

yongtang replied:
While trying to migrate some integration-cli tests to integration api tests, I noticed some of the tests rely on fixed "name" of containers, like `foo`, `server`, etc.. I think that may not work well with `parallel`. May need to convert them to random generated names to avoid collision.

dnephin replied:
Ya, I'm not against trying to get things running in parallel, but it is a lot more work than just adding `t.Parallel()`. 

It's a lot easier to get the cleanup wrong if we don't have an easy way to label everything with the test name.

thaJeztah replied:
> May need to convert them to random generated names to avoid collision.

Where possible we should just capture the container ID that's returned

cpuguy83 replied:
Yeah, I think we'd need to have the client automatically namespace objects or something then the defer could cleanup just the namespaces resources.

yongtang commented:
Overall LGTM. Though powerpc build keeps panicking. Not sure if it is related (or it exposes some other unrelated issue).

cpuguy83 commented on integration/container/kill_test.go:
```diff
@@ -0,0 +1,182 @@
+			err = client.ContainerKill(ctx, c.ID, "TERM")
+			require.NoError(t, err)
+
+			poll.WaitOn(t, containerIsInState(ctx, client, c.ID, tc.status), poll.WithDelay(100*time.Millisecond))
```
Just a performance concern here.. could this start up a "wait" goroutine earlier on for the case of it being stopped? (side thought, would be nice if the API let us wait for a container to be running).

cpuguy83 replied:
Funny, just found https://github.com/moby/moby/pull/36147

cpuguy83 reviewed (changes requested):
Seems mostly ok, the only thing is I don't think we can use `defer setupTest(t)()` with `t.Parallel()`... and I think we should opt for `t.Parallel()` over the cleanup helper.

vdemeester commented:
@cpuguy83 updated by removing `t.Parallel()` for now.

cpuguy83 reviewed (approved):
LGTM

yongtang commented:
@vdemeester I think there was an usage of `NewClientForHost` in secret_test.go which causes the build failure. Can you take a look?

vdemeester commented:
@yongtang yep, I will 😉

yongtang reviewed (approved):
LGTM
